### PR TITLE
Improve AesGcmSiv Aead performance by optimizing classpath check

### PR DESCRIPTION
**Pull request by ks-yim**
### Motivation:

`Aead` implementation for `AesGcmSiv` checks if `GCMParameterSpec` class exists in classpath by invoking `Class#forName` method whenever data gets encrypted or decrypted.

A benchmark on `AesGcmSiv` claims that `Class#forName` accounts for 4-5% of the total CPU cycles consumed for data en/decryption and 12% or more throughput gain can be obtained by optimizing away the `Class#forName` method invocation.

### Modification:

Cache the result of the classpath check in `AesGcmSiv`

### BM Results:

* [BM Code](https://github.com/ks-yim/tink-bm/blob/master/src/jmh/java/com/example/tink/bm/AesGcmSivBenchmark.java)
* Run on:
  - MacBook Pro (16-inch, 2021) Apple M1 Pro 32GB, MacOS Monterey version 12.6.2
  - Used `Conscrypt` custom build based on [3051cda](https://github.com/google/conscrypt/tree/3051cda0556eb142274d637443c7f5ae3b55a61f) to utilize AES-NI on Apple Silicon

* Before
  ```
  # JMH version: 1.35
  # VM version: JDK 11.0.17, OpenJDK 64-Bit Server VM, 11.0.17+8
  # VM invoker: /Users/ks-yim/.sdkman/candidates/java/11.0.17-tem/bin/java
  # VM options: -Dfile.encoding=UTF-8 -Djava.io.tmpdir=/Users/ks-yim/IdeaProjects/tink-bm/build/tmp/jmh - 
  Duser.country=KR -Duser.language=en -Duser.variant
  # Blackhole mode: full + dont-inline hint (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
  # Warmup: 1 iterations, 10 s each
  # Measurement: 1 iterations, 10 s each
  # Timeout: 10 min per iteration
  # Threads: 8 threads, will synchronize iterations
  # Benchmark mode: Throughput, ops/time
  # Benchmark: com.example.tink.bm.AesGcmSivBenchmark.conscrypt_encrypt
  
  Benchmark                                             Mode  Cnt       Score   Error  Units
  AesGcmSivBenchmark.bc_decrypt                        thrpt       140397.856          ops/s
  AesGcmSivBenchmark.bc_encrypt                        thrpt       161178.303          ops/s
  AesGcmSivBenchmark.conscrypt_decrypt                 thrpt       487751.715          ops/s
  AesGcmSivBenchmark.conscrypt_encrypt                 thrpt       462492.430          ops/s
  ```

* After
  ```
  Benchmark                                             Mode  Cnt       Score   Error  Units
  AesGcmSivBenchmark.bc_decrypt                        thrpt       168118.773          ops/s
  AesGcmSivBenchmark.bc_encrypt                        thrpt       172307.366          ops/s
  AesGcmSivBenchmark.conscrypt_decrypt                 thrpt       546720.978          ops/s
  AesGcmSivBenchmark.conscrypt_encrypt                 thrpt       551282.859          ops/s
  ```

### CPU Profiling Result:

* Encryption
  - Before
    flame-graph-cpu

  - After
     flame-graph-cpu

* Decryption
  - Before
     flame-graph-cpu

  - After
    flame-graph-cpu

### CODE REVIEW

File: src/main/java/com/google/crypto/tink/aead/subtle/AesGcmSiv.java [hunk 1]
```diff
@@ -58,6 +58,18 @@ protected Cipher initialValue() {
   private static final int IV_SIZE_IN_BYTES = 12;
   private static final int TAG_SIZE_IN_BYTES = 16;
```
Comment (tholenst) [inline src/main/java/com/google/crypto/tink/aead/subtle/AesGcmSiv.java:60]: I think this is slightly easier to understand if you just do:

private static final boolean HAS_GCM_PARAMETER_SPEC_CLASS = checkForGCMParameterSpec();

private static final boolean checkForGCMParameterSpec() {
    try {
      Class.forName("javax.crypto.spec.GCMParameterSpec");
    } catch (ClassNotFoundException e) {
       return false;
    }
  return true;
}


WDYT?

Comment (ks-yim) [reply src/main/java/com/google/crypto/tink/aead/subtle/AesGcmSiv.java:60]: Sorry for the late response.

No objection for this. I will send a new commit!

Comment (ks-yim) [reply src/main/java/com/google/crypto/tink/aead/subtle/AesGcmSiv.java]: Fixed it, PTAL 🙏.

**Comment by tholenst** [comment]
Thank you. I will try to merge this but still have to see if I figure out how we set this up. Please ping me if it didn't work.

**Comment by ks-yim** [comment]
@tholenst Thanks for reviewing this, but I didn't get you in the last sentence.
Should I try something and let you know the result if it doesn't work?

**Comment by tholenst** [comment]
Sorry for being unclear.

I am working on merging this, but the process works in some predefined way interacting with the Google-internal repository. I do not have experience with the process, so it might happen that it doesn't work and I forget about it.

It should be merged on Monday, if it didn't work I made a mistake and I would welcome a reminder.

**Comment by ks-yim** [comment]
@tholenst thanks for working on this, I will close the PR, then.

**Pull request closed without merging**